src/java.base/share/classes/sun/security/ssl/PreSharedKeyExtension.java

```diff
@@ -810,20 +834,27 @@ private static byte[] computeBinder(HandshakeContext context,
 
     private static SecretKey deriveBinderKey(HandshakeContext context,
             SecretKey psk, SSLSessionImpl session) throws IOException {
+        SecretKey earlySecret = null;
         try {
             CipherSuite.HashAlg hashAlg = session.getSuite().hashAlg;
-            HKDF hkdf = new HKDF(hashAlg.name);
+            KDF hkdf = KDF.getInstance(hashAlg.hkdfAlgorithm);
             byte[] zeros = new byte[hashAlg.hashLength];
-            SecretKey earlySecret = hkdf.extract(zeros, psk, "TlsEarlySecret");
-
+            earlySecret = hkdf.deriveKey("TlsEarlySecret",
+                    HKDFParameterSpec.ofExtract().addSalt(zeros).addIKM(psk)
+                    .extractOnly());
             byte[] label = ("tls13 res binder").getBytes();
             MessageDigest md = MessageDigest.getInstance(hashAlg.name);
             byte[] hkdfInfo = SSLSecretDerivation.createHkdfInfo(
                     label, md.digest(new byte[0]), hashAlg.hashLength);
-            return hkdf.expand(earlySecret,
-                    hkdfInfo, hashAlg.hashLength, "TlsBinderKey");
+            return hkdf.deriveKey("TlsBinderKey",
+                    HKDFParameterSpec.expandOnly(earlySecret, hkdfInfo,
+                    hashAlg.hashLength));
```

Comment:
Is it possible to combine the 2 `deriveKey` calls above into a single Extract-Then-Expand call? Then you don't need to clean up `earlySecret`.

Reply:
Should be possible, let me give it a try. Thanks~